src/highdicom/seg/content.py

```diff
@@ -152,6 +154,146 @@ def __init__(
+                AlgorithmIdentificationSequence.from_sequence(
+                    desc.SegmentationAlgorithmIdentificationSequence
+                )
+        if hasattr(desc, 'AnatomicRegionSequence'):
+            desc.AnatomicRegionSequence = [
+                CodedConcept.from_dataset(ds)
+                for ds in desc.AnatomicRegionSequence
+            ]
+        if hasattr(desc, 'PrimaryAnatomicStructureSequence'):
+            desc.PrimaryAnatomicStructureSequence = [
+                CodedConcept.from_dataset(ds)
+                for ds in desc.PrimaryAnatomicStructureSequence
+            ]
+        return desc
+
+    @property
+    def segment_number(self) -> int:
+        """int: Number of the segment."""
+        return self.SegmentNumber
+
+    @property
+    def segment_label(self) -> str:
+        """str: Label of the segment."""
+        return self.SegmentLabel
+
+    @property
+    def segmented_property_category(self) -> CodedConcept:
+        """highdicom.sr.coding.CodedConcept:
+            Category of the property the segment represents.
+
+        """
+        return self.SegmentedPropertyCategoryCodeSequence[0]
+
+    @property
+    def segmented_property_type(self) -> CodedConcept:
+        """highdicom.sr.coding.CodedConcept:
+            Type of the property the segment represents.
+
+        """
+        return self.SegmentedPropertyTypeCodeSequence[0]
+
+    @property
+    def algorithm_type(self) -> SegmentAlgorithmTypeValues:
+        """highdicom.seg.enum.SegmentAlgorithmTypeValues:
+            Type of algorithm used to create the segment.
+
+        """
+        return SegmentAlgorithmTypeValues(self.SegmentAlgorithmType)
+
+    @property
+    def algorithm_identification(
+        self
+    ) -> Optional[AlgorithmIdentificationSequence]:
+        """Optional[highdicom.content.AlgorithmIdentificationSequence]
+            Information useful for identification of the algorithm, if any.
+
+        """
+        if hasattr(self, 'SegmentationAlgorithmIdentificationSequence'):
+            return self.SegmentationAlgorithmIdentificationSequence
+        return None
+
+    @property
+    def tracking_uid(self) -> Optional[str]:
+        """Optional[str]:
+            Tracking unique identifier for the segment, if any.
+
+        """
+        if 'TrackingUID' in self:
+            return self.TrackingUID
+        return None
+
+    @property
+    def tracking_id(self) -> Optional[str]:
+        """Optional[str]: Tracking identifier for the segment, if any."""
+        if 'TrackingID' in self:
+            return self.TrackingID
+        return None
+
+    @property
+    def anatomic_regions(self) -> List[CodedConcept]:
+        """List[highdicom.sr.coding.CodedConcept]:
+            List of anatomic regions into which the segment falls.
+            May be empty.
+
+        """
+        if not hasattr(self, 'AnatomicRegionSequence'):
+            return []
+        return self.AnatomicRegionSequence
```

Comment:
We definitely need to be careful about this. However I think this "eager" type conversion is the right approach, and we should enforce via testing rather than runtime conversion in the parsing methods.

The reason is that I think that we should be careful to ensure that `from_dataset` methods convert all objects under their hierarchy to the relevant `highdicom` type where applicable so that objects returned from `from_dataset` are indistinguishable from those created via constructors and thereafter we don't have to worry about the "history" of the object. If we stick to this, it is clear where the burden lies for this type conversion (always in `from_dataset`) and I think this will lead to fewer bugs. I also discuss this on my [comment](https://github.com/MGHComputationalPathology/highdicom/pull/69#pullrequestreview-671889771) on the SR parsing API.

Reply:
Fully agreed. We should probably document this (and other fundamental design decision) in the [developer guide](https://highdicom.readthedocs.io/en/latest/development.html). What do you think?

Reply:
See #88